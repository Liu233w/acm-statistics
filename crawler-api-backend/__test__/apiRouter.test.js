/* eslint-disable no-undef */

// 在 require 的时候程序会直接 require 爬虫，如果在 beforeAll 里面 mock 的话就晚了
jest.mock('../../crawler/lib/configReader')
jest.mock('../../crawler/lib/functionGenerator')

const supertest = require('supertest')
const app = require('../app')

let server
let request

beforeAll(() => {
  server = app.listen(12001)
  request = supertest(server)
})

afterAll(() => {
  server.close()
})

test('/api/crawlers/swagger.json swagger should match snapshot', async () => {
  const res = await request.get('/api/crawlers/swagger.json')
  expect(res.status).toBe(200)
  expect(res.body).toMatchSnapshot()
})

test('/api/crawlers should return crawler list', async () => {
  const res = await request.get('/api/crawlers')
  expect(res.status).toBe(200)
  expect(res.body).toMatchObject({
    error: false,
    data: {
      crawler1: {
        title: 'Crawler1',
        description: 'Description1',
        url: 'http://www.c1.com',
      },
      crawler2: {
        title: 'Crawler2',
      },
      crawler_for_server: {
        title: 'CrawlerForServer',
      },
    },
  })
})

describe('/api/crawlers/:type/:username', () => {
  it('在结果正确时应该返回题量', async () => {
    const res = await request.get('/api/crawlers/crawler1/user')
    expect(res.status).toBe(200)
    expect(res.body).toMatchObject({
      error: false,
      data: {
        solved: 101,
        submissions: 230,
      },
    })
  })

  it('在爬虫不存在时应该返回 400', async () => {
    const res = await request.get('/api/crawlers/notExist/user')
    expect(res.status).toBe(400)
    expect(res.body).toMatchObject({
      error: true,
      message: '不存在此OJ的爬虫',
    })
  })

  it('在爬虫报错时应该返回错误结果', async () => {
    const res = await request.get('/api/crawlers/crawler1/reject')
    expect(res.status).toBe(400)
    expect(res.body).toMatchObject({
      error: true,
      message: '用户不存在',
    })
  })
})

test('在访问不存在的资源时返回 404', async () => {
  const res = await request.get('/notExists')
  expect(res.status).toBe(404)
  expect(res.body).toMatchObject({
    error: true,
    message: '404 Not Found',
  })
})