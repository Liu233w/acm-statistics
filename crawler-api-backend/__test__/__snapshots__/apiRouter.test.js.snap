// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/crawlers/swagger.json swagger should match snapshot 1`] = `
Object {
  "definitions": Object {
    "CrawlerMeta": Object {
      "description": "爬虫的元信息",
      "properties": Object {
        "description": Object {
          "description": "爬虫的说明",
          "type": "string",
        },
        "title": Object {
          "description": "爬虫的标题",
          "type": "string",
        },
        "url": Object {
          "description": "爬虫要爬的网址",
          "type": "string",
        },
      },
      "type": "object",
    },
    "Error": Object {
      "properties": Object {
        "error": Object {
          "description": "true",
          "type": "boolean",
        },
        "message": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "Ok": Object {
      "properties": Object {
        "data": Object {
          "$ref": "#/definitions/UserState",
        },
        "error": Object {
          "description": "false",
          "type": "boolean",
        },
      },
      "type": "object",
    },
    "UserState": Object {
      "description": "用户的题量状况",
      "properties": Object {
        "solved": Object {
          "format": "int64",
          "type": "integer",
        },
        "solvedList": Object {
          "description": "用户通过的题目列表。如果是普通的爬虫，只包含题目编号；如果是 virtual judge 的爬虫，格式为 \`crawlerName-problemNumber\`",
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
        "submissions": Object {
          "format": "int64",
          "type": "integer",
        },
        "submissionsByCrawlerName": Object {
          "additionalProperties": Object {
            "type": "number",
          },
          "default": null,
          "description": "If the crawler is virtual_judge, this field contains its submission count divided by crawler name.",
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "externalDocs": Object {
    "description": "后端文档",
    "url": "https://github.com/Liu233w/acm-statistics/blob/master/crawler-api-backend/README.md",
  },
  "info": Object {
    "description": "这是一个爬虫微服务。在这里可以通过 Rest API 调用 \`crawler\` 文件夹中的所有爬虫。 这些爬虫都是在服务器上运行的，因此您可以在任何地方使用本 API。",
    "title": "crawler-api-backend",
    "version": "1.0.0",
  },
  "paths": Object {
    "/api/crawlers/": Object {
      "get": Object {
        "consumes": Array [
          "application/json",
        ],
        "description": "",
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "success",
            "schema": Object {
              "properties": Object {
                "data": Object {
                  "additionalProperties": Object {
                    "$ref": "#/definitions/CrawlerMeta",
                  },
                  "description": "name of the crawler",
                  "type": "object",
                },
                "error": Object {
                  "description": "false",
                  "type": "boolean",
                },
              },
              "type": "object",
            },
          },
        },
        "summary": "返回所有支持查询的 OJ",
      },
    },
    "/api/crawlers/{type}/{username}": Object {
      "get": Object {
        "consumes": Array [
          "application/json",
        ],
        "parameters": Array [
          Object {
            "description": "OJ类型，只包括在 \`/crawler/config.yml\` 中定义的类型",
            "in": "path",
            "name": "type",
            "required": true,
            "type": "string",
          },
          Object {
            "in": "path",
            "name": "username",
            "required": true,
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "successful",
            "schema": Object {
              "$ref": "#/definitions/Ok",
            },
          },
          "400": Object {
            "description": "all other fails",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "summary": "查询某个OJ上的用户题量",
      },
    },
  },
  "schemes": Array [
    "http",
  ],
  "swagger": "2.0",
}
`;
